### DISCUSSION TITLE
User sessions

### DISCUSSION THREAD

**Discussion by 13hakta** [directvt/vtm#218] [General]
Sometimes when I reconnect to vtm labels like [1000:1] appears on screens.
Please describe how to manage sessions and what this label with session id means?

**Comment by o-sdn-o** [comment]
There is no global session management yet. Each user manages his session himself -- he can end his session with the `Disconnect` button. Also, each user can shut down all other sessions by pressing the `Shutdown` button - this will kick all users from the server and shut it down.
 
 `[1000:1]` is a combination of two values
  - `[1000:` - user ID issued by the operating system -- for example, in Linux for a regular user, the OS issues `1000`, MacOS issues `500`, Windows issues a login name string.
  - `:1]` is just the incremental number of the user connected to the server. This number simply increases by one each time the user connects, and never decreases.

**Reply by o-sdn-o** [reply]
In vtm, sessions are grouped using the user ID issued by the OS (e.g. `1000`). In other words, one vtm-server instance for one user ID. For different identifiers, individual servers are launched, and they do not overlap in any way.

**Reply by o-sdn-o** [reply]
vtm-server gets the ID of the connecting user by making a request to the operating system -- who connected to the pipe that the server is listening to and the operating system reply with such ID.

**Reply by o-sdn-o** [reply]
The following system calls are used to obtain the credentials of the connected client:
- Linux: [getsockopt(handle, SOL_SOCKET, SO_PEERCRED, &cred, &size)](https://man7.org/linux/man-pages/man7/socket.7.html)
- MacOS (BSD-like): [getpeereid(handle, &euid, &egid)](https://www.freebsd.org/cgi/man.cgi?query=getpeereid)
- Windows: [Named Pipes - default ACL used for a named pipe grant full control to the LocalSystem, admins, and the creator owner](https://docs.microsoft.com/en-us/windows/win32/ipc/named-pipes) - so no credentials needed.
